### DISCUSSION TITLE
Help for backend

### DISCUSSION THREAD

**Discussion by Kazilsky** [tailwindlabs/tailwindcss#13955] [Help] [answered]
I'm a backendr myself, I use node js, c#, c++, decided to learn frontend but decided to bend to the complicated side which I kinda almost handled, React, tailwind css, Vite, I use files with the jsx extension, how do I implement a login check on the main login page? 
There is a code sample of how it looks like on angular 
```
fetch("https://site/api/auth/login", {
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    },
    method: 'POST',
    body: JSON.stringify({'login':'user', 'password':'######'})
}).then(r => r.json()).then(x => console.log(x.data.token));
```
I understand the backend in general, but I don't understand how to take data from the "@material-tailwind/react" inputs when the button is pressed

**Comment by Sinabon2004** [comment, marked as answer]
maybe, like that?
```jsx
import React, { useState } from 'react';
import { Input, Button } from '@material-tailwind/react';

const LoginPage = () => {
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');

  const handleSubmit = async (e) => {
    e.preventDefault();
    const response = await fetch("https://site/api/auth/login", {
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body: JSON.stringify({ 'login': login, 'password': password })
    });
    const data = await response.json();
    console.log(data.token);
  };

  return (
    <form onSubmit={handleSubmit}>
      <Input 
        type="text" 
        placeholder="Username" 
        value={login} 
        onChange={(e) => setLogin(e.target.value)} 
      />
      <Input 
        type="password" 
        placeholder="Password" 
        value={password} 
        onChange={(e) => setPassword(e.target.value)} 
      />
      <Button type="submit">Login</Button>
    </form>
  );
};

export default LoginPage;

```
import React, { useState } from 'react';
import { Input, Button } from '@material-tailwind/react';

const LoginPage = () => {
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');

  const handleSubmit = async (e) => {
    e.preventDefault();
    const response = await fetch("https://site/api/auth/login", {
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body: JSON.stringify({ 'login': login, 'password': password })
    });
    const data = await response.json();
    console.log(data.token);
  };

  return (
    <form onSubmit={handleSubmit}>
      <Input 
        type="text" 
        placeholder="Username" 
        value={login} 
        onChange={(e) => setLogin(e.target.value)} 
      />
      <Input 
        type="password" 
        placeholder="Password" 
        value={password} 
        onChange={(e) => setPassword(e.target.value)} 
      />
      <Button type="submit">Login</Button>
    </form>
  );
};

export default LoginPage;

```

**Comment by VaiBrahma** [comment]
**It is quite simple actually.**

Here go through this code and read my explanation if you don't understand something.

### Note: You must install axios by  `npm install axios` in the root of your react directory where package.json is located.

```bash
import React, { useState } from 'react';
import { Input, Button } from '@material-tailwind/react';
import axios from 'axios';

const LoginPage = () => {
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const response = await axios.post('https://site/api/auth/login', { login, password }, { headers: { 'Accept': 'application/json', 'Content-Type': 'application/json' } });
      console.log(response.data.token);
    } catch (error) {
      console.error('Error logging in:', error);
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <Input type="text" placeholder="Username" value={login} onChange={(e) => setLogin(e.target.value)} />
      <Input type="password" placeholder="Password" value={password} onChange={(e) => setPassword(e.target.value)} />
      <Button type="submit">Login</Button>
    </form>
  );
};

export default LoginPage;
```

# explanation 

## return()

I have made a simple form which has two Input and one Button imported from `@material-tailwind/react`

### 1.Form 

This is a typical form which has an attribute `onSubmit={handleSubmit}`.

this attribute is crucial so that the `handleSubmit` function can fire everytime you click `Button` which has `type="submit"` attribute.

### 2.Input
Input has two importent attributes : 

1. *value*
2. *onChange*

value must have the **key** of your body and onChange you must change it's **value.** 

### 3.Button

This button must have an `type="submit"` attribute if you have given the form `onSubmit` attribute. You can also give `onClick={handleSubmit}` attribute to fire handleSubmit function when clicking the button but I don't recommend it because form attribute also enable you to submit form by clicking `enter` key on your keyboard. 

## Variables with useState

on the 6th, 7th line I have created two variables according to your body structure i.e. `login` and `password` , in the form you must provide them in the `value` attribute of `Input` and change their values by `onChange` attribute... Read about `useState` react hook to gain more insight about the function i wrote in the `onChange` attribute.

**Notice: that we only write `arrow function` or `function name` on `onChange`, `onSubmit` like attributes instead of calling these functions directly.**

## handleSubmit function

first we need to do e.preventDefault() and then I generally use `try and catch` block for api handling in react app.

then i did `axios.post(api, req, headers)`
1. api: your api string, 
2. req: your body object,
3. headers: necessary headers,

Do let me know if it still don't answer your question and if it answers the please **Mark this answer as the answer of your question.**

**Comment by Kazilsky** [comment]
Hello! In a month relatively well learned the frontend, thank you for your help earlier did not do, forgot to close the question but the next day seemed to be able to do it already, thank you for your help!
